cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -1203,6 +1203,47 @@ inline Status ConvertFixedSizeBinary(const ChunkedArray& data, PyObject** out_va
   return Status::OK();
 }
 
+inline Status ConvertStruct(const ChunkedArray& data, PyObject** out_values) {
+  PyAcquireGIL lock;
+  for (int c = 0; c < data.num_chunks(); c++) {
+    auto arr = static_cast<StructArray*>(data.chunk(c).get());
+    std::vector<PyObject*> fields_data;
```

Comment:
This will potentially make many small allocations on the heap, which will lead to memory fragmentation. One way to avoid that is to have a single vector on which you call `resize(arr->fields().size())` in each iteration of the outer loop and then assign to it e.g., `fields_data[i] = numpy_array;` in the first inner loop instead of appending with `push_back`.